### PR TITLE
[Snapping 3/5] Custom snapping API

### PR DESCRIPTION
**Pull request by SomeHats**
This diff adds an API for customising our existing snap types. These are:
1. Bound snapping. When translating or resizing a shape, it'll snap to certain key points on the bounds of particular shapes. Previously, these were hard-coded to the corners and center of the bounding box of the shape. Now, a shape can bring its own (e.g. a triangle may add snapping for its 3 corners, and it's centroid rather than bounding box center.
2. Handle outline snapping. When dragging a handle, it'll snap to the outline of other shapes geometry. Now, shapes can return different geometry for this sort of snapping if they like.

Each of these is customised through a method on `ShapeUtil`: `getBoundsSnapGeometry` and `getHandleSnapGeometry`. These return interfaces describing the different geometry that can be snapped to in both these cases. Currently, each returns an object with a single property, but there are more types of snapping coming in follow-up PRs. When reviewing this PR, start with the definitions of `BoundsSnapGeometry` in `BoundsSnaps.ts` and `HandleSnapGeometry` in `HandleSnaps.ts`

This doesn't add point snapping - i'll add that in a follow-up! It'll be customisable with the `getHandleSnapGeometry` API.

Fixes TLD-2197

This PR is part of a series - please don't merge it until the things before it have landed!
1. #2827 
4. #2831
5. #2793 (you are here)
6. #2841
7. #2845

### Change Type

- [x] `minor` — New feature

[^1]: publishes a `patch` release, for devDependencies use `internal`
[^2]: will not publish a new version

### Test Plan

- [x] Unit Tests

### Release Notes

- Add `ShapeUtil.getSnapInfo` for customising shape snaps.

### CODE REVIEW

**Review by ds300 — COMMENTED** [review]
Will finish a proper review tomorrow

File: packages/editor/src/lib/editor/managers/SnapManager/SnapManager.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import { atom, computed, EMPTY_ARRAY } from '@tldraw/state'
-import { isShapeId, TLFrameShape, TLGroupShape, TLParentId, TLShapeId } from '@tldraw/tlschema'
```
[the remark this answers is not in the document]
Comment (ds300) [reply packages/editor/src/lib/editor/managers/SnapManager/SnapManager.ts:2]: cc @si14

File: packages/tldraw/src/test/customSnapping.test.tsx [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+	}
+	const shapeUtils = [TestShapeUtil] as TLAnyShapeUtilConstructor[]
+
+	let editor: TestEditor
+	let ids: Record<string, TLShapeId>
+	beforeEach(() => {
+		editor = new TestEditor({ shapeUtils })
+		ids = editor.createShapesFromJsx([
+			<TL.geo ref="box" x={0} y={0} w={100} h={100} />,
+			<TL.test ref="test" x={200} y={200} w={100} h={100} boundsSnapPoints={null} />,
```
Comment (ds300) [inline packages/tldraw/src/test/customSnapping.test.tsx:45]: TIL we had this jsx thingy, nice!

**Review by ds300 — APPROVED** [review]
looks great, thanks for the thorough tests.

**Review by mimecuvalo — APPROVED** [review]
LGTM 👍

File: packages/editor/src/lib/editor/managers/SnapManager/SnapManager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (si14) [reply packages/editor/src/lib/editor/managers/SnapManager/SnapManager.ts:2]: I believe all CI changes were rolled back, see https://github.com/tldraw/tldraw/compare/7ad9ee0a67c9b66f547c428ae72580d9777b9139..b379a7a47c984e77c633f588f89c896aa2deded2 (that's one commit before biome and one commit after unbiome)

**Pull request merged**